covariance/covarianceXsec.cpp

```diff
@@ -94,59 +112,321 @@ covarianceXsec::covarianceXsec(const char *name, const char *file, double thresh
+	   }
+
+	   if (param["Systematic"]["NDSplineName"]) {
+		 _fNDSplineNames.push_back(param["Systematic"]["NDSplineName"].as<std::string>());
+	   }
+
+	 } else if(param["Systematic"]["Type"].as<std::string>() == "Norm") {
+	   std::cout << "Norm parameter" << std::endl;
+
+	   //First check to see if we have specified a mode
+	   std::vector<int> DummyModeVec;
+	   if(param["Systematic"]["Mode"]){
+		 _fNormModes.push_back(param["Systematic"]["Mode"].as<std::vector<int>>());
+	   } else{
+		 //Has to be of size 0 to mean apply to all
+		 _fNormModes.push_back(DummyModeVec);
+	   }
+	 }
+
+	 int NumKinematicCuts = 0;
+	 if(param["Systematic"]["KinematicCuts"]){
+
+	   NumKinematicCuts = param["Systematic"]["KinematicCuts"].size();
+	   std::cout << "Number of Kinematic cuts is " << NumKinematicCuts << std::endl;
+
+	   std::vector<std::string> TempKinematicStrings;
+	   std::vector<std::vector<double>> TempKinematicBounds;
+	   //First element of TempKinematicBounds is always -999, and size is then 3
+
+	   for(int KinVar_i = 0 ; KinVar_i < NumKinematicCuts ; ++KinVar_i){ 
+		 //ETA
+		 //This is a bit messy, Kinematic cuts is a list of maps
+		 //The for loop h
+		 for (YAML::const_iterator it=param["Systematic"]["KinematicCuts"][KinVar_i].begin();it!=param["Systematic"]["KinematicCuts"][KinVar_i].end();++it) {
+		   TempKinematicStrings.push_back(it->first.as<std::string>());
+		   TempKinematicBounds.push_back(it->second.as<std::vector<double>>());
+		   std::vector<double> bounds = it->second.as<std::vector<double>>();
+		 }
+	   }
+
+	   _fKinematicPars.at(i) = TempKinematicStrings;
+	   _fKinematicBounds.at(i) = TempKinematicBounds;	   
+	 }
+       
+	 //Also loop through the correlations
+	 if(param["Systematic"]["Correlations"]) {
+	   std::cout << "FOUND " << param["Systematic"]["Correlations"].size() << " CORRELATIONS!" << std::endl;
+	   for(int Corr_i = 0 ; Corr_i < param["Systematic"]["Correlations"].size()  ; ++Corr_i){
+		 for (YAML::const_iterator it=param["Systematic"]["Correlations"][Corr_i].begin();it!=param["Systematic"]["Correlations"][Corr_i].end();++it) {
+		   //std::cout << "Correlation with " << it->first.as<std::string>() << " of " << it->second.as<double>() << std::endl;
+           Correlations[i][it->first.as<std::string>()] = it->second.as<double>();
+		 }
+	   }
+	 }
+	 i++;
+  }
+  
+	 
+  //ETA
+  //Now that we've been through all systematic let's fill the covmatrix
+  //This makes the root TCov from YAML
+  for(int i=0; i < _fNumPar; i++) {
+	(*_fCovMatrix)(i,i)=_fError[i]*_fError[i];
+	//Get the map of parameter name to correlation fomr the Correlations object
+	for (auto const& [key, val] : Correlations[i]) {
+	  int index = -1;
+
+	  //If you found the parameter name then get the index
+	  if (CorrNamesMap.find(key) != CorrNamesMap.end()) {
+		index=CorrNamesMap[key];
+	  }
+	  else {
+		std::cout << "Parameter " << key << " not in list! Check your spelling?" << std::endl;
```

Comment:
We convert Yaml to ROOT matrix and then in CovBase ROOT to double** ;)
Which is fine for now